File: backend/app/schemas/backtest.py
from datetime import date

from pydantic import BaseModel

from app.schemas.stock import DataSourceMetadata


class BacktestPoint(BaseModel):
    date: date
    portfolio_value: float
    hold_value: float
    cumulative_return: float
    hold_return: float
    signal: str


class TradeRecord(BaseModel):
    entry_date: date
    exit_date: date | None
    entry_price: float
    exit_price: float | None
    return_pct: float | None


class StrategyBacktestSummary(BaseModel):
    strategy: str
    total_return: float
    max_drawdown: float
    win_rate: float
    trade_count: int
    average_trade_return: float
    sharpe_ratio: float | None = None


class BacktestResponse(DataSourceMetadata):
    ticker: str
    period: str
    strategy: str = "regime_adjusted_strategy"
    initial_capital: float
    total_return: float
    max_drawdown: float
    sharpe_ratio: float | None = None
    win_rate: float
    trade_count: int
    average_trade_return: float
    buy_and_hold_return: float
    legacy_strategy_return: float | None = None
    new_strategy_return: float | None = None
    strategy_results: list[StrategyBacktestSummary] = []
    chart: list[BacktestPoint]
    trades: list[TradeRecord]
    note: str

File: frontend/lib/types/backtest.ts
/** 백테스트 결과와 전략별 요약. */

import type { BacktestStrategy, DataSourceMetadata } from "./common";

export interface BacktestPoint {
  date: string;
  portfolio_value: number;
  hold_value: number;
  cumulative_return: number;
  hold_return: number;
  signal: string;
}

export interface TradeRecord {
  entry_date: string;
  exit_date: string | null;
  entry_price: number;
  exit_price: number | null;
  return_pct: number | null;
}

export interface BacktestResponse extends DataSourceMetadata {
  ticker: string;
  period: string;
  strategy: BacktestStrategy | string;
  initial_capital: number;
  total_return: number;
  max_drawdown: number;
  sharpe_ratio?: number | null;
  win_rate: number;
  trade_count: number;
  average_trade_return: number;
  buy_and_hold_return: number;
  legacy_strategy_return?: number | null;
  new_strategy_return?: number | null;
  strategy_results: StrategyBacktestSummary[];
  chart: BacktestPoint[];
  trades: TradeRecord[];
  note: string;
}

export interface StrategyBacktestSummary {
  strategy: string;
  total_return: number;
  max_drawdown: number;
  win_rate: number;
  trade_count: number;
  average_trade_return: number;
  sharpe_ratio?: number | null;
}
